game-state-service

```diff
@@ -0,0 +1,123 @@
+        GObject.Object.__init__(self)
+        self._state = {}
+        self._monitor = None
+        self._monitor_changed_id = None
+        self._monitor_restart_id = None
+        self._monitor_start()
+        self._state_from_file()
+
+    def _get_state_file(self):
+        path = GLib.build_filenamev([GLib.get_user_data_dir(), 'state.json'])
+        return Gio.File.new_for_path(path)
+
+    def _monitor_start(self):
+        state_file = self._get_state_file()
+        self._monitor = state_file.monitor_file(
+            Gio.FileMonitorFlags.NONE,
+            None)
+        self._monitor.set_rate_limit(self._MONITOR_RATE)
+        self._monitor_changed_id = self._monitor.connect(
+            'changed',
+            self._monitor_change_cb)
+
+    def _monitor_change_cb(self, monitor, child, other_file, event):
+        if event == Gio.FileMonitorEvent.CHANGES_DONE_HINT:
+            self._state_from_file()
+
+    def _monitor_restart_cb(self):
+        self._monitor_restart_id = None
+        self._monitor.handler_unblock(self._monitor_changed_id)
+        return False
+
+    def _state_from_file(self):
+        state_file = self._get_state_file()
+        try:
+            success, state_data, tag = state_file.load_contents(None)
+            self._state = json.loads(state_data.decode())
+        except:
+            pass
+
+    def _state_to_file(self):
+        state_data = json.dumps(self._state,
+                                sort_keys=True,
+                                indent=2,
+                                separators=(',', ': '))
+        state_file = self._get_state_file()
+
+        # always block the monitor when writing to the state file
+        if self._monitor_restart_id is None:
+            self._monitor.handler_block(self._monitor_changed_id)
+        else:
+            GLib.source_remove(self._monitor_restart_id)
+
+        try:
+            state_file.replace_contents(
+                bytes(state_data.encode()),
+                None,
+                False,
+                0,
+                None)
+        except:
+            pass
+        finally:
+            self._monitor_restart_id = GLib.timeout_add(
+                self._MONITOR_RATE * 2,
+                self._monitor_restart_cb)
+
+    def get(self, key):
+        if key not in self._state:
+            raise NotFoundException()
+        return self._state[key]
+
+    def set(self, key, value):
+        self._state[key] = value
+        self._state_to_file()
+
+
+class GameStateService(dbus.service.Object):
+
+    def __init__(self, bus, path):
+        dbus.service.Object.__init__(self, bus, path)
+        self._state = GameState()
+
+    @dbus.service.method(DBUS_IFACE, in_signature='s', out_signature='v')
+    def Get(self, key):
+        return self._state.get(key)
+
+    @dbus.service.method(DBUS_IFACE, in_signature='sv', out_signature='')
+    def Set(self, key, value):
+        return self._state.set(key, value)
```

Comment:
What is the interface we're using? Is it documented anywhere? Do we not need a signal to notify when the state has changed?

Reply:
The iface definition is explicit now, since I am using GDBus.

I think we could have a signal to let clients know that the state changed (globally), but I rather not try to mimic GObject properties-like API (for specific keys) or something like that for now.